src/managers/cluster.py

```diff
@@ -26,3 +37,93 @@ def get_host_mapping(self) -> dict[str, str]:
+        leader = ""
+
+        # loop through list of hosts and compare their member id with the leader
+        # if they match, return this host's endpoint
+        for endpoint in self.cluster_endpoints:
+            client = EtcdClient(client_url=endpoint)
+            try:
+                endpoint_status = client.get_endpoint_status()
+                member_id = endpoint_status["Status"]["header"]["member_id"]
+                leader_id = endpoint_status["Status"]["leader"]
+                if member_id == leader_id:
+                    leader = endpoint
+                    break
+            except KeyError:
+                # for now, we don't raise an error if there is no leader
+                # this may change when we have actual relevant tasks performed against the leader
+                logger.warning("No raft leader found in cluster.")
+
+        return leader
+
+
+class EtcdClient:
+    """Handle etcd client connections and run etcdctl commands."""
+
+    def __init__(
+        self,
+        client_url: str,
+    ):
+        self.client_url = client_url
+
+    def get_endpoint_status(self) -> dict:
+        """Run the `endpoint status` command and return the result as dict."""
+        endpoint_status = {}
+        if result := self._run_etcdctl(
+            command="endpoint",
+            subcommand="status",
+            endpoints=self.client_url,
+            output_format="json",
+        ):
+            endpoint_status = json.loads(result)[0]
+
+        return endpoint_status
+
+    def _run_etcdctl(
+        self,
+        command: str,
+        subcommand: str | None,
+        endpoints: str,
+        output_format: str | None,
+    ) -> str:
+        """Execute `etcdctl` command via subprocess.
+
+        The list of arguments will be extended once authentication/encryption is implemented.
+        This method aims to provide a very clear interface for executing `etcdctl` and minimize
+        the margin of error on cluster operations.
+
+        Args:
+            command: command to execute with etcdctl, e.g. `elect`, `member` or `endpoint`
+            subcommand: subcommand to add to the previous command, e.g. `add` or `status`
+            endpoints: str-formatted list of endpoints to run the command against
+            output_format: set the output format (fields, json, protobuf, simple, table)
+            ...
+
+        Returns:
+            The output of the subprocess-command as a string. In case of error, this will
+            return an empty string. It will not raise an error in order to leave error handling
+            up to the caller. Depending on what command is executed, the ways of handling errors
+            might differ.
+        """
+        try:
+            result = subprocess.run(
+                args=[
+                    "etcdctl",
+                    command,
+                    subcommand,
```

Comment:
There is also the [`check_output`](https://docs.python.org/3/library/subprocess.html#subprocess.check_output) which basically does this for you.

Reply:
I've started the implementation with `check_output`, but as the recommended approach to invoking subprocesses is to use the `run()` function, I switched to that.

I also tried to avoid "passing a list of arguments". I know this is possible and very convenient, but I've also seen a lot of issues with this. A recent one was masking credentials from being written to logfiles, which is much harder if you have just a list instead of dedicated parameters for e.g. password. Therefore I opted for a "controlled" approach of explicitly creating and naming those parameters that we need. This list will have to be extended eventually, that's true. Let's try to follow the approach of adding complexity only then when we need it. That means adding e.g. the arguments for the subcommands when we encounter the first need for it.